service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentPersonMapper.java

```diff
@@ -17,40 +17,58 @@
+            .agentId(agentDirectoryId);
 
-        buildPractitionerPrefix(practitioner).ifPresent(builder::practitionerPrefix);
-        var practitionerGiven = buildPractitionerGivenName(practitioner);
-        var practitionerFamily = buildPractitionerFamilyName(practitioner);
+        Optional<Practitioner> practitioner = messageContext.getInputBundleHolder()
+            .getResource(new IdType(agentKey.getPractitionerReference()))
+            .map(resource -> (Practitioner) resource);
 
-        practitionerGiven.ifPresent(builder::practitionerGivenName);
-        practitionerFamily.ifPresent(builder::practitionerFamilyName);
+        if (practitioner.isPresent()) {
+            builder.practitioner(true);
+            buildPractitionerPrefix(practitioner.get()).ifPresent(builder::practitionerPrefix);
 
-        if (practitionerGiven.isEmpty() && practitionerFamily.isEmpty()) {
-            builder.practitionerFamilyName(UNKNOWN);
+            var practitionerGiven = buildPractitionerGivenName(practitioner.get());
+            var practitionerFamily = buildPractitionerFamilyName(practitioner.get());
+
+            practitionerGiven.ifPresent(builder::practitionerGivenName);
+            practitionerFamily.ifPresent(builder::practitionerFamilyName);
+
+            if (practitionerGiven.isEmpty() && practitionerFamily.isEmpty()) {
+                builder.practitionerFamilyName(UNKNOWN);
+            }
         }
 
-        practitionerRole.flatMap(this::buildPractitionerRole).ifPresent(builder::practitionerRole);
-        organization.map(organizationToAgentMapper::mapOrganizationToAgentInner).ifPresent(builder::organization);
+        buildPractitionerRole(agentKey).ifPresent(builder::practitionerRole);
+
+        messageContext.getInputBundleHolder()
+            .getResource(new IdType(agentKey.getOrganizationReference()))
+            .map(resource -> (Organization) resource)
+            .map(OrganizationToAgentMapper::mapOrganizationToAgentInner)
+            .ifPresent(builder::organization);
 
         return TemplateUtils.fillTemplate(AGENT_STATEMENT_TEMPLATE, builder.build());
     }
 
-    private Optional<String> buildPractitionerRole(PractitionerRole practitionerRole) {
-        if (hasPractitionerRoleDisplay(practitionerRole)) {
-            return Optional.of(practitionerRole.getCodeFirstRep().getCodingFirstRep().getDisplay());
+    private Optional<String> buildPractitionerRole(AgentDirectory.AgentKey agentKey) {
+        Optional<PractitionerRole> practitionerRole = messageContext.getInputBundleHolder()
+            .getPractitionerRoleFor(
+                agentKey.getPractitionerReference(),
+                agentKey.getOrganizationReference()
+            );
+
+        if (practitionerRole.isPresent() && hasPractitionerRoleDisplay(practitionerRole.get())) {
+            return Optional.of(practitionerRole.get().getCodeFirstRep().getCodingFirstRep().getDisplay());
         }
+
         return Optional.empty();
```

Comment:
This is equivalent to the following 
```suggestion
        return practitionerRole
            .filter(AgentPersonMapper::hasPractitionerRoleDisplay)
            .map(PractitionerRole::getCodeFirstRep)
            .map(CodeableConcept::getCodingFirstRep)
            .map(Coding::getDisplay);
```

If you think it's more readable